File: nextpermutation.java
public class Solution {
    
    public void reverse(int[] num, int left, int right){
        while(left<right){
            int tmp = num[left];
            num[left] = num[right];
            num[right] = tmp;
            left++;
            right--;
        }
    }
    
    
    public void nextPermutation(int[] num) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int last = num.length-1;
        while(true){
            if(last==0){
                break;
            }
            if(num[last]<=num[last-1]){
                last--;
            }else{
                break;
            }
        }
        if(last==0){
            reverse(num,0,num.length-1);
            return;
        }
        int start = last-1;
        int i;
        for(i=last;i<num.length;i++){
            if(num[i]<=num[start]){
                break;
            }
        }
        int tmp = num[start];
        num[start] = num[i-1];
        num[i-1] = tmp;
        reverse(num,last,num.length-1);
        return;
    }
}

File: cpp/NextPermutation.cpp
class Solution {
public:
    void nextPermutation(vector<int> &num) {
        if(num.size()<=1) {
            return;
        }
        int firstDescending = 0;
        for(int i=num.size()-1;i>0;i--) {
            if(num[i]>num[i-1]) {
                int val = num[i-1];
                int j=i;
                while(j<num.size()) {
                    if(num[j] > val) {
                        j++;   
                    } else {
                        break;
                    }
                }
                swap(num,j-1,i-1);
                reverseV(num, i, num.size()-1);
                return;
            }
        }
        reverseV(num, 0, num.size()-1);
    }
    
    void reverseV(vector<int>& num, int start, int end) {
        while(start<end) {
            swap(num, start, end);
            start++;
            end--;
        }
    }
    
    void swap(vector<int>& num, int indexa, int indexb) {
        int temp = num[indexa];
        num[indexa] = num[indexb];
        num[indexb] = temp;
    }
};
